File: cpp/alias.cpp
#include <iostream>
#include <vector>
using namespace std;

template <typename T> void
make_room(vector<T> &v, size_t requirement)
{
  size_t newSize = max(requirement, v.size());
  v.reserve(newSize);
  v.resize(newSize);
}

template <typename T> void
array_copy(vector<T> &src, vector<T> &dest,
           size_t srcOff, size_t destOff,
           size_t len)
{
  make_room(dest, destOff + len);
  for(unsigned int i = 0; i < len; ++i)
    dest[destOff + i] = src[srcOff + i];
}

int main(int, char **)
{
  auto v = vector<int> {1, 2, 3};
  array_copy(v, v, 0, 1, 3);

  for(auto i : v) { cout << i << '\n'; }
}

File: rs/src/alias.rs
#[cfg(feature="never")]
mod will_fail {
    use std::cmp::max;

    unsafe fn make_room<T>(v : &mut Vec<T>, req : usize) {
        let new_size = max(req, v.len());
        v.reserve(new_size);
        v.set_len(new_size);
    }

    fn array_copy<T : Clone>(
        src  : &Vec<T>,
        dest : &mut Vec<T>,

        src_off  : usize,
        dest_off : usize,

        len : usize) {
        let iter =
            src[src_off..src_off+len]
              .iter().enumerate();

        unsafe {
            make_room(dest, dest_off + len);
            for (i, x) in iter {
                dest.insert(dest_off + i, x.clone());
            }
        }
    }

    fn test() {
        let mut v = vec![1, 2, 3];
        array_copy(&mut v, &mut v, 0, 1, 3);
    }
}
